File: 2405. Optimal Partition of String/list_2.py
class Solution:
    def partitionString(self, s: str) -> int:
        partitions = 1 # start with 1 as all characters might be unique
        last_seen_list = [-1]*26
        partition_start = 0
        for index in range(len(s)):
            if last_seen_list[ord(s[index]) - ord('a')] >= partition_start :
                partitions += 1
                partition_start = index
            last_seen_list[ord(s[index]) - ord('a')] = index
        return partitions
    
# time O(n)
# space O(26)

File: 2405. Optimal Partition of String/list_2.cpp
# include <vector>
# include <string>

class Solution {
public:
    int partitionString(string s) {
        int partitions = 1;
        vector<int> last_seen_list(26, -1);
        int partition_start = 0;
        for (int index = 0; index < s.length(); index++){
            if (last_seen_list[s[index] - 'a'] >= partition_start){
                partitions += 1;
                partition_start = index;
            }
            last_seen_list[s[index] - 'a'] = index;
        }
        return partitions;
    }
};